Add ability to format query for editor in client

azat opened the pull request:
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add ability to format query for editor in client (via M-F)

The formatting will multi-line if the query is multi-line itself, and
online otherwise.

You can use EDITOR=cat to just format queries in the client instead of
editing them.

Note, that the old shortcut (M-E) has not been replaced, and will not
format query, since sometimes in case of queries that does not fit into
a screen you don't want this

Screencast - https://asciinema.org/a/Xx7qk73bgXj2bg7OIrEZB5Ejm

v2: use C-; over C-E (it is readline jump to end of line)
v3: use M-F (F stands for format) over M-E (C-; is tricky since it does not have normal representation as C-[a-z] has)

azat committed:
client: repaint the client on error

Before:

    :) Empty query

After:

    :) Empty query
    :)

azat committed:
client: add ability to format query for editor (via M-F)

The formatting will multi-line if the query is multi-line itself, and
online otherwise.

You can use EDITOR=cat to just format queries in the client instead of
editing them.

Note, that the old shortcut (M-E) has not been replaced, and will not
format query, since sometimes in case of queries that does not fit into
a screen you don't want this

v2: use C-; over C-E (it is readline jump to end of line)
v3: use M-F (F stands for format) over M-E (C-; is tricky since it does not have normal representation as C-[a-z] has)

azat committed:
client: add one more line before printing error

Before:

    :) asdSyntax error (Query): failed at position 1 (asd): asd. ...
    :) asd

After:

    :) asd
    Syntax error (Query): failed at position 1 (asd): asd. ...
    :) asd

azat committed:
tests: cover formatting query for editor in client

P.S. I don't see how I can send Alt-Enter with expect, so I cannot
emulate the multiple queries, each on new line

azat committed:
client: fix re-render after query editing in editor

azat committed:
Move query formatting into client code out from ReplxxLineReader

Michicosun commented on src/Client/ClientBaseHelpers.cpp:
```diff
@@ -237,4 +238,25 @@ void highlight(const String & query, std::vector<replxx::Replxx::Color> & colors
+{
+    bool multiline_query = query.contains('\n');
+    ParserQuery parser(query.data() + query.size(), /*allow_settings_after_format_in_insert_=*/ false, /*implicit_select_=*/ false);
+    const ASTPtr ast = parseQuery(parser, query, /*max_query_size=*/ 0, /*max_parser_depth=*/ 0, /*max_parser_backtracks=*/ 0);
```
This will probably throw an exception on multi-statement query. Let's use `parseQueryAndMovePosition` and format each of the parsed queries independently?

```
:) select 1; select 2; select 10;
Syntax error (. Multi-statements are not allowed): failed at position 9 (end of query): ; select 2; select 10;.
```

azat replied:
Good catch, fixed

Michicosun commented on src/Client/ReplxxLineReader.cpp:
```diff
@@ -530,12 +534,17 @@ void ReplxxLineReader::addToHistory(const String & line)
+        String query = rx.get_state().text();
+
+        if (format_query)
+            query = formatQuery(query);
```
```suggestion
            query = formatQuery(std::move(query));
```

Michicosun commented on src/Client/ClientBaseHelpers.cpp:
```diff
@@ -237,4 +238,25 @@ void highlight(const String & query, std::vector<replxx::Replxx::Color> & colors
+    ParserQuery parser(query.data() + query.size(), /*allow_settings_after_format_in_insert_=*/ false, /*implicit_select_=*/ false);
+    const ASTPtr ast = parseQuery(parser, query, /*max_query_size=*/ 0, /*max_parser_depth=*/ 0, /*max_parser_backtracks=*/ 0);
+
+    bool insert_with_data = false;
+    if (auto * insert_query = ast->as<ASTInsertQuery>(); insert_query && insert_query->hasInlinedData())
+        insert_with_data = true;
```
Why not to simply format this? If the query was parsed then it should be able to format into string with the inlined data I suppose, otherwise it should throw an exception in parser

azat committed:
Properly format multiple queries

azat committed:
Add ability to format the INSERT with VALUES/FORMAT

azat committed:
Even better query substitution in replxx

Michicosun commented on src/Client/ClientBaseHelpers.cpp:
```diff
@@ -237,4 +241,53 @@ void highlight(const String & query, std::vector<replxx::Replxx::Color> & colors
+    {
+        const char * query_start = pos;
+
+        Tokens tokens(pos, end);
+        IParser::Pos token_iterator(tokens, max_parser_depth, max_parser_backtracks);
+        while (token_iterator->type != TokenType::Semicolon && token_iterator.isValid())
+            ++token_iterator;
+        const char * query_end = token_iterator->end;
+        bool has_semicolon = token_iterator.isValid() && token_iterator->type == TokenType::Semicolon;
```
Why do you need this? 

I think this approach https://github.com/ClickHouse/ClickHouse/blob/6b938d60f82fe2ff3f73b16cff8254fefe24eef1/src/Common/Scheduler/Workload/WorkloadEntityKeeperStorage.cpp#L248 will be suitable.

azat replied:
It was needed to detect does query has semicolon or not, but it can be simplified. Done.

azat committed:
Forbid multi-line queries during formatting query for editor

We already detect the end of the query above, so if we will allow
multi-line queries detected query end and real query end may not match

azat committed:
Merge u/master into client-editor-format

azat committed:
Move skipSpacesAndComments() into client helpers

azat committed:
Handle INSERT Values properly while query formatting for editor

Michicosun commented on src/Client/ClientBaseHelpers.cpp:
```diff
@@ -237,4 +243,110 @@ void highlight(const String & query, std::vector<replxx::Replxx::Color> & colors
+                pos = this_query_end;
+
+                /// Remove semicolon from the INSERT query payload, since it will be added explicitly below
+                if (*insert_ast->end == '\n')
```
Don't we need to check something like `insert_ast->end < end` before dereferencing?

azat replied:
Not really since it is a NUL terminated string.
